sympy/solvers/solveset.py

```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+
+    * Proof of correctness of the method
+
+    The logarithm is the inverse function to exponentiation.
+    The defining relation between exponentiation and logarithm is:
+
+    .. math:: {log_b x} = y \enspace if \enspace b^y = x
+
+    Therefore if we are given an equation with exponent terms, we can
+    convert every term to its corresponding log form. This is achieved by
+    taking logarithms and expanding it using log identities so that the
+    equation can be easily handled by `solveset`.
+
+    For example:
+
+    .. math:: 3^{2x} = 2^{x + 3}
+
+    Taking log both sides will reduce the equation to
+
+    .. math:: (2x)*log(3) = (x + 3)*log(2)
+
+    This form can be easily handed by `solveset`.
+    """
+
+    if not (isinstance(f, Add) and len(f.args) == 2):
+        # solving for the sum of more than two powers is possible
+        # but not yet implemented
+        return None
+
+    lhs, rhs = list(ordered(f.args))
+
+    log_type_equation = expand_log(log(lhs)) - expand_log(log(-rhs))
+
+    return log_type_equation
+
+
+def _check_expo(f, symbol):
+    r"""
+    Helper to check whether an equation is exponential or not.
+
+    ``f`` is the equation to be checked
+
+    ``symbol`` is the variable in which the equation needs to be checked
+
+    The function extracts each term of the equation and checks if it is
+    of exponential form w.r.t `symbol`.
+    Returns `True` if any term is of exponential type otherwise `False`.
+
+    Examples
+    ========
+
+    >>> from sympy import symbols, cos, exp
+    >>> from sympy.solvers.solveset import _check_expo as check
+    >>> x, y = symbols('x y')
+    >>> check(x**y - x, y)
+    True
+    >>> check(x**y - x, x)
+    False
+    >>> check(exp(x + 3), x)
+    True
+    >>> check(cos(2**x), x)
+    False
+    """
+
+    for add_arg in Add.make_args(f):
+        for mul_arg in Mul.make_args(add_arg):
+            if isinstance(mul_arg, (Pow, exp)) and \
+                    symbol in mul_arg.exp.free_symbols:
+                return True
+    return False
+
+
+def _transolve(f, symbol, domain):
+    r"""
+    Function to solve transcendental equations. It is a helper to
+    `solveset` and should be used internally as of now. `\_transolve`
+    currently supports the following class of equations:
+
+        - Exponential equations
+
+    Parameters
+    ==========
+
+    f : can be any transcendental equation that needs to be solved.
+        This needs to be an expression, which is assumed
+        to be equal to 0.
+
+    symbol : is variable for which the equation is solved.
+        This needs to be of class `Symbol`.
+
+    domain : is a set over which the equation is solved.
+        This needs to be of class `Set`.
+
+    Examples
+    ========
+
+    >>> from sympy.solvers.solveset import _transolve as transolve
+    >>> from sympy.solvers.solvers import _tsolve as tsolve
+    >>> from sympy import symbols, S, pprint
+    >>> x = symbols('x', real=True)
```

Comment:
In keeping with solveset, the results should not depend on the symbol assumptions. If you need to make the symbol real, you can do so internally.

Reply:
This is the danger of the docstring example, I suppose: making it look like transolve can/should be called independently of solveset. Perhaps a note clarifying this interaction would be in order. (Or indicate at the outset of the examples that although transolve is being called directly here, normally it would be called by solveset that would take care of putting the expression in a canonical form and passing the relevant assumptions on the symbols.

(But if assumptions are passed on the symbol, then why is the domain necessary? Perhaps any domain clipping should be done back in solveset. But take this comment noting that I am not yet thoroughly familiar with the workings of transolve.)